### DISCUSSION TITLE
DataTable cell text style setting!

### DISCUSSION THREAD

**Discussion by Yurii-huang** [Textualize/textual#4043] [Q&A] [answered]
```python
from textual.events import Key
from textual.widgets import DataTable
from rich.style import Style

class Options(DataTable):

    ROWS = [
        ("Options", "Stauts"),
        ("Configure the network", "no stting"),
        ("Set up users and password", "no stting"),
        ("Clock from NTP server", "no stting"),
        ("Disks partition", "no stting"),
        ("Set up bootloader", "no stting"),
        ("Choose operation after installed", "no stting"),
        ("Start the installation", ">>>>>>")
    ]

    ROWS_TEXT_STYLE = [
        (Style(), Style()),
        (Style(), Style(color="red", bold=True)),
        (Style(), Style(color="red", bold=True)),
        (Style(), Style(color="red", bold=True)),
        (Style(), Style(color="red", bold=True)),
        (Style(), Style(color="red", bold=True)),
        (Style(), Style(color="red", bold=True)),
        (Style(), Style(color="green", bold=True))
    ]
    
    def __init__(self):
        super().__init__()
        self.add_columns(*self.ROWS[0])
        self.add_rows(self.ROWS[1:])
        self.show_header = True
        self.cursor_type = "row"
        
    def _get_row_renderables(self, row_index: int):
        ret = super()._get_row_renderables(row_index)
        if self.show_header: # has row -1 offset
            row_index += 1
        styles = self.ROWS_TEXT_STYLE[row_index]
        for col in range(0, len(ret.cells)):
            rich_text = ret.cells[col]
            rich_text.style = styles[col]
            ret.cells[col] = rich_text
        return ret
```

**Comment by Yurii-huang** [comment]
Try changing the ROWS_TEXT_STYLE to dynamically change the Text Style

**Comment by TomJGooding** [comment, marked as answer]
There's a section about [Styling and justifying cells](https://textual.textualize.io/widgets/data_table/#styling-and-justifying-cells) in the DataTable docs. You really don't need to be messing with internal functions at all, just use a `Text` object.

**Reply by Yurii-huang** [reply]
Thank you for your answer. 
I guess I made a mistake with the program entry. 
I always want to implement these classes through inheritance. 
Instead of using it directly, and I need high customization support, I am still learning Textual

**Reply by davep** [reply]
The thing to be aware of is that internals, by their very nature, may change their behaviour, be renamed or disappear from release to release. Building on top of internals like that could cause you problems.

More generally, perhaps you could be a bit more clear with your questions? Your question was just some code that asked nothing; it would be a lot easier for people to help you if you were clear as to what your question is.

**Reply by Yurii-huang** [reply]
You reminded me that for Python's standard syntax, functions starting with underscores mean that the class is private, which may change in the release, and there seems to be no mention of the proposed inheritance functions in the documentation, so can I think that functions other than the standard syntax defines underscore functions can be inherited?

**Reply by Yurii-huang** [reply]
I'll give you an example, this can show that the default component may not be applicable, the default height of the Input component is 3, which contains styles such as boder, which is very beautiful, but in some scenarios, for example, I only have the serial output interface of the machine, and the height of my serial interface is not very much, this kind of component with a height of 3 will take up more space, resulting in an incomplete display, so I want to reduce the height of the native Input component, which should be difficult to achieve (I tried to change to height 1, The cursor, characters, will not be displayed). This may be my application scenario is relatively simple, it is a display interface for an operating system installation

**Reply by Yurii-huang** [reply]
Thank every commitor

**Reply by TomJGooding** [reply]
> I want to reduce the height of the native Input component, which should be difficult to achieve (I tried to change to height 1, The cursor, characters, will not be displayed).

There's an [Additional Notes](https://textual.textualize.io/widgets/input/#additional-notes) section in the `Input` docs which you might have missed, as you'll also need to remove the border:

```python
from textual.app import App, ComposeResult
from textual.widgets import Input


class SingleLineInputApp(App):
    CSS = """
    Input {
        height: 1;
        border: none;
    }

    Input:focus {
        border: none;
    }
    """

    def compose(self) -> ComposeResult:
        yield Input()


if __name__ == "__main__":
    app = SingleLineInputApp()
    app.run()
```

We're not saying that you shouldn't create your own custom widgets (far from it!) so of course you could create your own subclass of Input if you'd prefer.  But we wouldnt recommend accessing of internals (the methods with the leading underscore) for all the reasons Dave mentioned, especially when most of the time you don't need to!

```python
from textual.app import App, ComposeResult
from textual.widgets import Input


class SingleLineInput(Input):
    DEFAULT_CSS = """
    SingleLineInput {
        height: 1;
        border: none;
    }

    SingleLineInput:focus {
        border: none;
    }
    """


class SingleLineInputApp(App):
    def compose(self) -> ComposeResult:
        yield SingleLineInput()


if __name__ == "__main__":
    app = SingleLineInputApp()
    app.run()
```

**Reply by Yurii-huang** [reply]
Gad, You are right.
You are very professional in terms of component familiarity. As I said, I am a beginner, both in terms of documentation and familiarity with components. I now know that the correct solution to a problem should describe what kind of effect I want to achieve, rather than achieving it through inexplicable means.
